packages/context/src/watcher.ts

```diff
@@ -0,0 +1,255 @@
+import {
+  findAppRoot,
+  getProjectRoot,
+  getProjectInfo,
+  getPackageName,
+} from "./utils";
+import {
+  findManifestPath,
+  getMainService,
+  getManifestDetails,
+  getUI5Manifest,
+} from "./manifest";
+import { getProject, getApp } from "./loader";
+import { cache } from "./cache";
+import { Manifest } from "@sap-ux/project-access";
+import { FileChangeType, FileEvent } from "vscode-languageserver/node";
+import { URI } from "vscode-uri";
+import { getYamlDetails } from "./ui5-yaml";
+import { join } from "path";
+import { FileName } from "@sap-ux/project-access";
+import { getLogger } from "@ui5-language-assistant/logic-utils";
+
+const packageName = getPackageName();
+
+/**
+ * React on manifest.json file change
+ *
+ * @param uri uri to manifest.json file
+ * @description
+ * a. remove manifest cache
+ *
+ * b. remove app cache
+ *
+ * c. remove project cache
+ *
+ * d. get a fresh project
+ */
+export const reactOnManifestChange = async (
+  manifestUri: string,
+  changeType: FileChangeType
+): Promise<void> => {
+  getLogger(packageName).debug("`reactOnManifestChange` function called", {
+    manifestUri,
+    changeType,
+  });
+  const manifestPath = URI.parse(manifestUri).fsPath;
```

Comment:
The idea is to have cache on function level which may have performance bottleneck like `getCAPServices` Though it is related to CAP project but handle individually this is because CAP can have multiple apps where services can be same.  This avoid double compilation. Maybe in the feature we can think of path being hierarchical or similar approach and give away performance over code readability. But that is a matter of a different PR.